core/src/main/scala/filodb.core/downsample/DownsamplePublisher.scala

```diff
@@ -0,0 +1,53 @@
+package filodb.core.downsample
+
+import scala.concurrent.Future
+
+import com.typesafe.scalalogging.StrictLogging
+
+import filodb.core.{Response, Success}
+
+/**
+  * Dispatches downsample data to the FiloDB datasets holding downsampled data.
+  */
+trait DownsamplePublisher {
+  /**
+    * Start the downsample publish pipeline/thread
+    */
+  def start(): Unit
+
+  /**
+    * Cleanly stop the downsample publish task. Typically called on shutdown.
+    */
+  def stop(): Unit
+
+  /**
+    * Dispatch samples to the downsampling dataset.
+    * The publisher needs to take care of retry logic, and acks if any.
+    * Flush pipeline in TimeSeriesShard is expected to move on and not hold off the ingestion pipeline
+    * if there is any slowdown here.
+    *
+    * For now, there is a possibility of loss of samples in memory that have not been dispatched.
+    *
+    * @param shardNum
+    * @param resolution
+    * @param records each Array of Byte is a bunch of Binary Records that can be deserialized by the ingestion pipeline.
+    * @return Future of Success if all is good.
+    */
+  def publish(shardNum: Int, resolution: Int, records: Seq[Array[Byte]]): Future[Response]
```

Comment:
I haven't gotten to how this is tied into the flush logic, but it seems to me that you want an Observable.  The original  flush chunks is an Observable, it seems you can then feed that into downsampling logic and emit and Observable of containers.  See how I do it in the GatewayServer.

Reply:
See Step 3 - documentation just added in the doFlushSteps of TimeSeriesShard for details on how this is linked to the flush pipeline.

The kafka publish is indeed an observable inside the Kafka publisher, but it is not connected to the flush observable because of the checkpointing.

Today, I have queued things in memory for better throughput, but we may decide to change our mind and wait for an ack from Kafka before checkpointing.